src/lib/config.ts

```diff
@@ -0,0 +1,417 @@
+    if (v !== null && typeof v === "object") {
+      return new Config(this.keyName, v as any);
+    } else {
+      throw new InvalidConfigError(
+        `Expected object for ${this.keyName} ${key}, got ${typeof v}`
+      );
+    }
+  }
+
+  str(key: string): string {
+    return this.require(key, this.strOpt(key));
+  }
+
+  number(key: string): number {
+    return this.require(key, this.numberOpt(key));
+  }
+
+  require<T>(key: string, value: T | undefined): T {
+    if (value === undefined) {
+      throw new InvalidConfigError(
+        `Missing value for required ${this.keyName} ${key}`
+      );
+    } else {
+      return value;
+    }
+  }
+
+  entries(pred: (key: string) => boolean): [string, Config][] {
+    return Object.keys(this.config).flatMap((k) =>
+      pred(k) ? [[k, this.object(k)]] : []
+    );
+  }
+}
+
+// Represents `~/.fauna-shell`
+class RootConfig {
+  defaultEndpoint?: string;
+  endpoints: { [key: string]: Endpoint };
+
+  constructor(config: Config) {
+    this.defaultEndpoint = config.strOpt("default");
+    this.endpoints = Object.fromEntries(
+      config
+        .entries((k) => k !== "default")
+        .map(([k, v]) => [k, new Endpoint(v)])
+    );
+
+    if (this.defaultEndpoint === "default") {
+      throw new InvalidConfigError(
+        "Default endpoint cannot be named 'default'"
+      );
+    } else if (
+      this.defaultEndpoint &&
+      this.endpoints[this.defaultEndpoint] === undefined
+    ) {
+      throw new InvalidConfigError(
+        `Default endpoint '${this.defaultEndpoint}' was not found`
+      );
+    }
+  }
+}
+
+// Represents `.fauna-project`
+class LocalConfig {
+  defaultStack: string | undefined;
+  stacks: { [key: string]: Stack };
+
+  constructor(config: Config) {
+    this.defaultStack = config.strOpt("default");
+    this.stacks = Object.fromEntries(
+      config.entries((k) => k !== "default").map(([k, v]) => [k, new Stack(v)])
+    );
+
+    if (this.defaultStack === "default") {
+      throw new InvalidConfigError("Default stack cannot be named 'default'");
+    } else if (
+      this.defaultStack &&
+      this.stacks[this.defaultStack] === undefined
+    ) {
+      throw new InvalidConfigError(
+        `Default stack '${this.defaultStack}' was not found`
+      );
+    }
+  }
+
+  validate(rootConfig: RootConfig) {
+    for (const stack of Object.values(this.stacks)) {
+      if (rootConfig.endpoints[stack.endpoint] === undefined) {
+        throw new InvalidConfigError(
+          `Endpoint '${stack.endpoint}' not found in ~/.fauna-shell`
+        );
+      }
+    }
+  }
+}
+
+class Endpoint {
+  secret: string;
+  url?: string;
+
+  graphqlHost?: string;
+  graphqlPort?: number;
+
+  constructor(config: Config) {
+    this.secret = config.str("secret");
+    this.url = getURLFromConfig(config);
+
+    this.graphqlHost = config.strOpt("graphqlHost");
+    this.graphqlPort = config.numberOpt("graphqlPort");
+  }
+
+  makeEndpoint(
```

Comment:
wondering if we could call this something like:
`withScopeOrRole` and then just return a new Endpoint object hat is updated accordingly.

Reply:
the endpoint can update the url and graphql host/port, so I'm not sure that name would work